src/Tools/dotnet-monitor/CollectionRules/CollectionRuleService.cs

```diff
@@ -232,5 +242,96 @@ private async Task StopAndDisposeContainerAsync(CancellationToken token)
+            CollectionRulePipeline.DequeueOldTimestamps(pipeline.ExecutionTimestamps, limitsOptions?.ActionCountSlidingWindowDuration, currentTime);
+
+            int actionCountLimit = (limitsOptions?.ActionCount).GetValueOrDefault(CollectionRuleLimitsOptionsDefaults.ActionCount);
+            pipeline.CheckForThrottling(actionCountLimit, pipeline.ExecutionTimestamps.Count, limitsOptions?.ActionCountSlidingWindowDuration);
+
+            var description = new CollectionRuleDescription()
+            {
+                State = pipeline.StateHolder.CurrState,
+                StateReason = pipeline.StateHolder.CurrStateReason,
+                LifetimeOccurrences = pipeline.AllExecutionTimestamps.Count,
+                ActionCountLimit = actionCountLimit,
+                SlidingWindowOccurrences = pipeline.ExecutionTimestamps.Count,
+                ActionCountSlidingWindowDurationLimit = limitsOptions?.ActionCountSlidingWindowDuration,
+            };
+
+            if (description.State != CollectionRuleState.Finished)
+            {
+                description.SlidingWindowDurationCountdown = GetSWDCountdown(pipeline.ExecutionTimestamps, description.ActionCountSlidingWindowDurationLimit, description.ActionCountLimit, currentTime);
+                description.RuleFinishedCountdown = GetRuleFinishedCountdown(pipeline.PipelineStartTime, limitsOptions?.RuleDuration, currentTime);
+            }
+
+            return description;
+        }
+
+        private static TimeSpan? GetSWDCountdown(Queue<DateTime> timestamps, TimeSpan? actionCountWindowDuration, int actionCount, DateTime currentTimestamp)
+        {
+            if (actionCountWindowDuration.HasValue && timestamps.Count >= actionCount)
+            {
+                TimeSpan countdown =  timestamps.Peek() - (currentTimestamp - actionCountWindowDuration.Value);
+                return GetTruncatedPositiveTimeSpan(countdown);
+            }
+
+            return null;
+        }
+
+        private static TimeSpan? GetRuleFinishedCountdown(DateTime pipelineStartTime, TimeSpan? ruleDuration, DateTime currentTimestamp)
+        {
+            if (ruleDuration.HasValue)
+            {
+                TimeSpan countdown = ruleDuration.Value - (currentTimestamp - pipelineStartTime);
+                return GetTruncatedPositiveTimeSpan(countdown);
+            }
+
+            return null;
+        }
+
+        private static TimeSpan? GetTruncatedPositiveTimeSpan(TimeSpan original)
+        {
+            return (original > TimeSpan.Zero) ? TimeSpan.FromSeconds((long)original.TotalSeconds) : null; // Intentionally lose millisecond precision
+        }
+
+        private void CleanUpCompletedPipelines(IEndpointInfo key)
+        {
+            var collectionRuleNamesFrequency = new Dictionary<string, int>();
+
+            var activePipelineTimestamps = new Dictionary<string, DateTime>();
+
+            foreach (var pipeline in _containersMap[key]._pipelines)
```

Comment:
I think it would be better if we did this using eventing or something similar. For example, once a pipeline ends we callback into the service (or provide an event on the pipeline itself to indicate that it has ended) and remove the pipeline as necessary.

Reply:
Look into this more - should be able to trash all of our pipelines since we will be resetting everything (move it to when configuration changes) -> will clear out everything that's finished from a previous run.